src/loaders/FileLoader.js

```diff
@@ -55,225 +53,178 @@ class FileLoader extends Loader {
 
-						break;
+				if ( response.status === 200 || response.status === 0 ) {
 
-					case 'document':
-
-						const parser = new DOMParser();
-						response = parser.parseFromString( data, mimeType );
-
-						break;
-
-					case 'json':
-
-						response = JSON.parse( data );
-
-						break;
-
-					default: // 'text' or other
-
-						response = data;
+					// Some browsers return HTTP Status 0 when using non-http protocol
+					// e.g. 'file://' or 'data://'. Handle as success.
 
-						break;
+					if ( response.status === 0 ) {
 
-				}
+						console.warn( 'THREE.FileLoader: HTTP Status 0 received.' );
 
-				// Wait for next browser tick like standard XMLHttpRequest event dispatching does
-				setTimeout( function () {
+					}
 
-					if ( onLoad ) onLoad( response );
+					const callbacks = loading[ url ];
+					const reader = response.body.getReader();
+					const contentLength = response.headers.get( 'Content-Length' );
+					const total = contentLength ? parseInt( contentLength ) : 0;
+					const lengthComputable = total != 0;
+					let loaded = 0;
 
-					scope.manager.itemEnd( url );
+					// periodically read data into the new stream tracking while download progress
+					return new ReadableStream( {
+						start( controller ) {
 
-				}, 0 );
+							readData();
 
-			} catch ( error ) {
+							function readData() {
 
-				// Wait for next browser tick like standard XMLHttpRequest event dispatching does
-				setTimeout( function () {
+								reader.read()
+									.then( ( { done, value } ) => {
 
-					if ( onError ) onError( error );
+										if ( done ) {
 
-					scope.manager.itemError( url );
-					scope.manager.itemEnd( url );
+											controller.close();
 
-				}, 0 );
+										} else {
 
-			}
+											loaded += value.byteLength;
 
-		} else {
+											const event = new ProgressEvent( 'progress', { lengthComputable, loaded, total } );
+											for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
-			// Initialise array for duplicate requests
+												const callback = callbacks[ i ];
+												if ( callback.onProgress ) callback.onProgress( event );
 
-			loading[ url ] = [];
+											}
 
-			loading[ url ].push( {
-
-				onLoad: onLoad,
-				onProgress: onProgress,
-				onError: onError
+											controller.enqueue( value );
+											readData();
 
-			} );
+										}
 
-			request = new XMLHttpRequest();
+									} );
 
-			request.open( 'GET', url, true );
+							}
 
-			request.addEventListener( 'load', function ( event ) {
+						}
+        	} );
 
-				const response = this.response;
+				} else {
 
-				const callbacks = loading[ url ];
+					const callbacks = loading[ url ];
+					delete loading[ url ];
 
-				delete loading[ url ];
+					for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
-				if ( this.status === 200 || this.status === 0 ) {
+						const callback = callbacks[ i ];
+						if ( callback.onError ) callback.onError( response.statusText );
```

Comment:
I think it might be best to throw an Error here in the case that there's an issue because otherwise the promise will fall through to the next `stream =>` callback with an "undefined" value causing a _second_ error to be thrown and caught anyway. Throwing an error with the status text here will also let us remove redundant on error callback code.

Reply:
that's a good idea. on it.